### PR TITLE
GEODE-6854: Skipped events already contained in the batch during conf…

### PR DESCRIPTION
**Pull request by boglesby**
…lation

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java [hunk 1]
```diff
@@ -1179,6 +1190,38 @@ public Long getShadowKey() {
     return this.shadowKey;
   }
 
+  public boolean equals(Object obj) {
+    if (this == obj) {
+      return true;
+    }
+
+    if (obj == null || !(obj instanceof GatewaySenderEventImpl)) {
```
Comment (mcmellawatt) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: I think the null check is redundant here

File: geode-core/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderEventProcessorJUnitTest.java [hunk 2]
```diff
@@ -151,4 +151,69 @@ public void validateBatchConflationWithBatchContainingDuplicateNonConflatableEve
     // Assert no events were conflated incorrectly
     assertThat(originalEvents).isEqualTo(conflatedEvents);
   }
+
+  @Test
+  public void validateBatchConflationWithBatchContainingDuplicateConflatableAndNonConflatableEvents()
+      throws Exception {
+    // This is a test for GEODE-6854.
```
Comment (mcmellawatt) [inline geode-core/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderEventProcessorJUnitTest.java]: I think we might want to avoid using ticket numbers in comments if we can.  If we ever change ticketing systems, this can make it hard to track down the history of why the changes were made.  A short description of the problem is more useful IMO.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java (hunk 1 above)
Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: Thanks for the review. I'll remove that check and also add a unit test for this method.

File: geode-core/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderEventProcessorJUnitTest.java (hunk 2 above)
Comment (boglesby) [reply geode-core/src/test/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderEventProcessorJUnitTest.java]: You're right about this. I'll remove the JIRA and update the description.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java [hunk 3]
```diff
@@ -774,6 +776,15 @@ public String toString() {
     return buffer.toString();
   }
 
+  public String toSmallString() {
+    StringBuffer buffer = new StringBuffer();
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: I think you want to use 'StringBuilder' here to avoid synchronization overhead from 'StringBuffer'.

File: geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImplTest.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.geode.internal.cache.wan.parallel.ParallelGatewaySenderHelper;
+import org.apache.geode.test.fake.Fakes;
+import org.apache.geode.test.junit.rules.serializable.SerializableTestName;
+
+public class GatewaySenderEventImplTest {
+
+  private GemFireCacheImpl cache;
+
+  @Rule
+  public SerializableTestName testName = new SerializableTestName();
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImplTest.java]: You can just use `org.junit.rules.TestName` here. The serializable version is really just for dunit tests where you want to read the test name from any VM.

File: geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImplTest.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  @Test
+  public void testEquality() throws Exception {
+    LocalRegion region = mock(LocalRegion.class);
+    when(region.getFullPath()).thenReturn(testName.getMethodName() + "_region");
+    when(region.getCache()).thenReturn(cache);
+    Object event = ParallelGatewaySenderHelper.createGatewaySenderEvent(region, Operation.CREATE,
+        "key1", "value1", 0, 0, 0, 0);
+
+    // Basic equality tests
+    assertThat(event.equals(null)).isFalse();
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImplTest.java]: I think you should unwind we assertions to:
```
assertThat(event).isNotNull();
assertThat(event).isEqualTo(event);
```
...and so on. Looks like maybe you used a conversion script from junit to assertj or something.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java (hunk 3 above)
Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: Thanks for the review. I meant to use StringBuilder. I guess I picked the wrong auto-completion.

File: geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImplTest.java (hunk 4 above)
Comment (boglesby) [reply geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImplTest.java]: Thanks. I'll fix this.

File: geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImplTest.java (hunk 5 above)
Comment (boglesby) [reply geode-core/src/test/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImplTest.java]: Thanks. I'll fix this. I'm not sure I knew isEqualTo existed.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java (hunk 3 above)
Comment (boglesby) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: Oh, I see what I did here. I just copied the existing toString method and modified it. I'll fix that method too.
